src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -147,20 +151,21 @@ public SolutionState(
 
     /// <summary>
     /// A list of all the ids for all the projects contained by the solution.
+    /// Ordering determined by the order the projects were added to the solution.
     /// </summary>
     public IReadOnlyList<ProjectId> ProjectIds { get; }
 
     private void CheckInvariants()
     {
         // Run these quick checks all the time.  We need to know immediately if we violate these.
-        Contract.ThrowIfFalse(ProjectStates.Count == ProjectIds.Count);
-        Contract.ThrowIfFalse(ProjectStates.Count == _dependencyGraph.ProjectIds.Count);
+        Contract.ThrowIfFalse(ProjectStates.Length == ProjectIds.Count);
+        Contract.ThrowIfFalse(ProjectStates.Length == _dependencyGraph.ProjectIds.Count);
 
         // Only run this in debug builds; even the .SetEquals() call across all projects can be expensive when there's a lot of them.
 #if DEBUG
         // project ids must be the same:
-        Debug.Assert(ProjectStates.Keys.SetEquals(ProjectIds));
-        Debug.Assert(ProjectStates.Keys.SetEquals(_dependencyGraph.ProjectIds));
+        Debug.Assert(ProjectStates.Select(static state => state.Id).SetEquals(ProjectIds));
+        Debug.Assert(ProjectStates.Select(static state => state.Id).SetEquals(_dependencyGraph.ProjectIds));
```

Comment:
Want to add another assertion that ProjectStates is in order?

Reply:
Done!